regparser/tree/xml_parser/paragraph_processor.py

```diff
@@ -58,10 +55,23 @@ def build_hierarchy(self, root, nodes, depths):
                 # Note that nodes still only have the one label part
                 label, cnt = self.clean_label(node.label[0], cnt)
                 node.label = [label]
+                # Children should also have labels applied to them:
+                if len(node.children):
+                    node = self.label_child_nodes(node)
                 stack.add(1 + par.depth, node)
 
         return stack.collapse()
 
+    def label_child_nodes(self, node):
+        """
+        Takes a node and recursively processes its children to add the
+        appropriate labels to them.
+        """
+        for idx, child in enumerate(node.children):
+            child.label = node.label + ["p%s" % str(idx + 1)]
```

Comment:
This works so long as we don't care about the inner children's original marker. There might be a way to use `clean_label` (which only replaces `MARKERLESS`), but we can come to that bridge later.